spec/consensus/gossip/crdt.md

```diff
@@ -68,59 +62,64 @@ To save memory, entries should be removed from the tuple space as soon as they b
 For example, suppose a node's local view of the tuple space has the following entries, here organized as rows of a table for easier visualization:
 
-| Height | Round | Step      | Validator | Payload |
-| ------ | ----- | --------- | --------- | ------- |
-| 1      | 0     | Proposal  | v1        | pp      |
-| 1      | 0     | PreVote   | v1        | vp      |
-| 1      | 1     | PreCommit | v2        | cp      |
-| 2      | 0     | Proposal  | v1        | pp'     |
-| 2      | 2     | PreVote   | v2        | vp'     |
-| 2      | 3     | PreCommit | v2        | cp'     |
+| Height | Round | Step      | Validator | Payload        |
+| ------ | ----- | --------- | --------- | -------------- |
+| 1      | 0     | Proposal  | v1        | pp1            |
+| 1      | 0     | PreVote   | v1        | vp1            |
+| 1      | 1     | PreCommit | v2        | cp1            |
+| 2      | 0     | Proposal  | v1        | pp2            |
+| 2      | 2     | PreVote   | v2        | vp2            |
+| 2      | 3     | PreCommit | v2        | cp2   [^todo2] |
+| 2      | 3     | PreCommit | v2        | cp2'  [^todo2] |
 
-- Query $\lang 0, 0, Proposal, v1, * \rang$ returns $\{ \lang 0, 0, Proposal, v1, pp \rang \}$
-- Query $\lang 0, 0, *, v1, * \rang$ returns $\{ \lang 0, 0, Proposal, v1, pp \rang,  \lang 0, 0, PreVote, v1, vp \rang \}$.
+- Query $\lang 0, 0, Proposal, v1, * \rang$ returns $\{ \lang 0, 0, Proposal, v1, pp1 \rang \}$
+- Query $\lang 0, 0, *, v1, * \rang$ returns $\{ \lang 0, 0, Proposal, v1, pp1 \rang,  \lang 0, 0, PreVote, v1, vp1 \rang \}$.
 
 If needed for disambiguation, queries are subscripted with the node being queried.
 
+[^todo2]: These tuples are evidence of an equivocation attack. It is not clear yet if we should keep both entries in the local view.
+
 #### State Validity
 
-Let $\text{ValSet}_h \subseteq P$ be the set of validators of height $h$ and $\text{Prop}_{h,r}$ be the proposer of round $r$ of height $h$.
+Let $V_h \subseteq P$ be the set of validators of height $h$ and $\pi^h_r \in V_h$ be the proposer of round $r$ of height $h$.
+When the context eliminates any ambiguity on the height number, we might write these values simply as $V$ and $\pi_r$.
 
-Given that each validator can execute each step only once per round, a query that specifies height, round, step and validator must either return empty or a single tuple.
+Given that each validator can execute each step only once per round, a query that specifies height, round, step and validator SHOULD either return empty or a single tuple.
 
-- $\forall h \in \N, r \in \N, s \in \{\text{Proposal, PreVote, PreCommit}\}, v \in \text{ValSet}_h$,  $\cup_{p \in P} \lang h, r, s, v, * \rang_p$ contains at most one element.
+- $\forall h \in \N, \forall r \in \N, \forall s \in \{\text{Proposal, PreVote, PreCommit}\}, \forall v \in V_h$,  $\cup_{p \in P} \lang h, r, s, v, * \rang_p$ contains at most one element.
```

Comment:
I do not understand $\cup{p \in P} (h,r,s,v,*)$ contains at most one element.

In general, is it not better to have these preconditions on the tuple spaces and consider that they are enforced at all time?

Reply:
I didn't understand your comment.
What I tried to express here is that there should be a single vote per validator per (height, round).
Before the vote is issued, the query result is empty. If the validator behaves, eventually there may be one entry in the set. But if the validator misbehaves, there could be more than one value in the result.

Reply:
I see. $(h,r,s,v,*)$ is supposed to be a query, whose result is a set. I will improve on the notation.

You are absolutely right about more than one message existing here being related to that discussion.